8264737: JavaFX media stream stops playing after reconnecting via Remote Desktop

sashamatveev opened the pull request:
Remote Desktop will change default audio device when connecting or disconnecting to it's own audio device. Also, when remote desktop is disconnect, then remote desktop default audio device is removed and system default device is not restored until user logs back to computer. So, after remote desktop is disconnected, then system left without any default audio devices.

To fix this we will unload DirectSound when audio device is gone and will continue playback by throwing away audio data. Once we receive notification that default audio device is back, we will load it and continue playback via this device.

Loading device done in loop, since it is not always available right after notification, but will be after very short period of time.

Tested by connecting/disconnecting remote desktop and switching between remote and normal desktop. In second case audio will switch between remote or speakers.

Since audio device is gone and not restored after disconnect, there will be no audio on machine local speakers, until user logs back again.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264737](https://bugs.openjdk.java.net/browse/JDK-8264737): JavaFX media stream stops playing after reconnecting via Remote Desktop


### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/479/head:pull/479` \
`$ git checkout pull/479`

Update a local copy of the PR: \
`$ git checkout pull/479` \
`$ git pull https://git.openjdk.java.net/jfx pull/479/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 479`

View PR using the GUI difftool: \
`$ git pr show -t 479`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/479.diff">https://git.openjdk.java.net/jfx/pull/479.diff</a>

</details>

kevinrushforth commented:
/reviewers 2

kevinrushforth reviewed (commented):
As far as I can tell, the code changes look fine to me.

While testing this, the stream always keeps playing now. Sometimes I can't hear the audio, when it reconnects while playing the stream the first time. What I did was:

1. Run Ensemble
2. Select the "Audio Bar Chart" sample
3. Disconnect from remote desktop
4. Reconnect to remote desktop

The audio stream continues to play, as seen by the chart continuing to be updated, but I can't hear anything. If I select any other sample, and then select the Audio Bar Chart sample again, I can disconnect / reconnect and everything is fine -- I can hear the audio.

Btw, the same thing happens if I initially select the Advanced Media sample. If it is the first media sample that is run, then when I disconnect / reconnect, I can't hear anything, but the video keeps playing.

kevinrushforth commented:
As a related data point, even if I don't disconnect while playing the initial media stream, if I do the following, there is no problem and I can hear the audio on reconnect:

1. Run Ensemble
2. Select the "Audio Area Chart"
3. Select the "Audio Bar Chart" sample
4. Disconnect from remote desktop
5. Reconnect to remote desktop

No problem.

sashamatveev commented:
Audio device notification failed in case when CoInitialize() was not called. It worked in some cases, since other parts of JavaFX already initialized it. Fixed by adding CoInitialize() call before registering device notification. Also, removed unused header left from debug printf.

kevinrushforth reviewed (approved):
Looks good.

arapte commented:
The change fixes the issue and looks good.
While testing it regressively, I observed an issue. The issue can occur always.
steps to reproduce:
1. Run Ensemble app on remote desktop
2. Run Streaming Media Player app
3. Press the media player buttons for several times. Keep pressing different buttons for a minute or more and at the end press stop backward.
4. Disconnect remote desktop and connect again
5. Press play button
6. I observed two different issues at two different instances
- Video and audio bot do not play
- Audio plays but video does not play

However, Looking at the steps to reproduce this does not seem like a stopper to me, so approving
Let me know if you can or cannot reproduce the issue, we can take that as a follow on issue.

sashamatveev commented:
@arapte I was not able to reproduce issue as described. It might not be related to removal of audio device during Remote Desktop disconnect. I will go ahead and integrate fix. Can you file a bug for it?

sashamatveev commented:
/integrate
